### PR TITLE
[python] Append mode with resizing

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** As tracked on issue #2407 / [[sc-51048]](https://app.shortcut.com/tiledb-inc/story/51048).

Note that the intended Python and R API changes are all agreed on and finalized as described in #2407.

**Changes:**

Support experipment-level `upgrade_shape` and `resize`.

**Notes for Reviewer:**

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -1203,9 +1206,13 @@ def _write_dataframe_impl(
         arrow_table = _extract_new_values_for_append(df_uri, arrow_table, context)
 
     try:
+        domain = None
+        if NEW_SHAPE_FEATURE_FLAG_ENABLED:
+            domain = ((0, shape - 1),)
```
Comment (jp-dark) [inline apis/python/src/tiledbsoma/io/ingest.py:1211]: It looks like this is now assuming the dataframe uses the `soma_joinid` as the one and only index column. If this is the correct assumption (and I think it probably is for the anndata ingestor), can you add a comment about that in the docstring?

**Review by jp-dark — APPROVED** [review]
I think this looks good, but I'm not super familiar with the ingestor code so you may want to want to wait for @nguyenv to review as well.

One question: Is that resize code required before ingesting with the new shape feature? If so, what error does the user get if they fail to resize first?

File: apis/python/src/tiledbsoma/io/ingest.py (hunk 1 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:1211]: I'll do so

Yes `tiledbsoma.io` does choose `soma_joinid` as the one and only index column.

**Comment by johnkerl** [comment]
> One question: Is that resize code required before ingesting with the new shape feature? If so, what error does the user get if they fail to resize first?

@jp-dark they get a message like this:

```
tiledbsoma._exception.SOMAError: TileDB internal: ['UnorderedWriter::dowork]  (FragmentMetadata: Cells are written outside of the defined current domain.)
```

which we can perhaps catch and rethrow with a friendlier error message ...

**Review by nguyenv — APPROVED** [review]

**Pull request merged**